src/stemmabench/variant_analyzer.py

```diff
@@ -0,0 +1,703 @@
+"""
+This module analyzes an alignment table of variant texts and provides 
+estimates of the frequency and distribution of various types of transformation 
+(modification operations) within the overall tradition.
+"""
+
+# Import libraries.
+import numpy as np
+import textdistance
+
+from loguru import logger
+from typing import List, Dict
+from itertools import combinations
+from collatex.core_classes import AlignmentTable
+import nltk; nltk.download("wordnet")
+from nltk.corpus import wordnet
+from nltk.stem import WordNetLemmatizer
+
+from stemmabench.data import SUPPORTED_LANGUAGES
+
+# Define VariantAnalyzer class.
+class VariantAnalyzer:
+
+    def __init__(self,
+                 table: AlignmentTable,
+                 language: str = "en",
+                 disable_synonym: bool = False) -> None:
+        """_summary_
+
+        Args:
+            table (AlignmentTable): _description_.
+            array (Union[np.ndarray[object], None], optional): _description_. 
+                Defaults to None.
+            language (str, optional): _description_. 
+                Defaults to "en".
+        """
+        # alignment table
+        self.table = table
+        self.variant_locations: List[bool] = [col.variant for col in table.columns] 
+        # language
+        if language not in SUPPORTED_LANGUAGES:
+            logger.critical(f"Unknown language {language}.")
+            raise ValueError(f"Unknown language {language}.")
+        self.language = language
+        self.disable_synonym = disable_synonym
+
+    @property
+    def witness_names(self) -> List[str]:
+        """Get the list of witness names from an AlignmentTable.
+
+        Returns:
+            List[str]: The list of witness names extracted from the table 
+                rows.
+        """
+        return [row.header for row in self.table.rows]
+    
+    @property
+    def array(self) -> np.ndarray[np.ndarray]:
+        """Get an array representing an input Collatex AlignmentTable. 
+
+        Returns:
+            np.ndarray: A 2D numpy array representing the alignment table.
+                - Rows are witnesses.
+                - Columns are readings.
+        """
+        return self.alignment_table_to_numpy(self.table)
+    
+    #------------------------------------------------------------------------
+    # ------------ UTILITY METHODS ------------------------------------------
+    #------------------------------------------------------------------------
+    @staticmethod
+    def _get_token_strings(tokens_list: List[str | None], missing_reading:str="-"
```

Comment:
missing extra space in arguments
```suggestion
    def _get_token_strings(tokens_list: List[str | None], missing_reading: str="-"
```

Reply:
- [x] add extra space in arguments